src/Tools/DynamoInstallDetective/ProductLookUp.cs

```diff
@@ -0,0 +1,345 @@
+        {
+            if (string.IsNullOrEmpty(basePath))
+                return false;
+
+            return Directory.Exists(basePath) && Directory.GetFiles(basePath, fileLookup).Any();
+        }
+
+        public virtual string GetInstallLocationFromProductName(string name)
+        {
+            var key = OpenKey(REG_KEY64 + name);
+            return GetInstallLocation(key);
+        }
+
+        public virtual string GetInstallLocationFromProductCode(string productCode, out string productName)
+        {
+            string issProdKey = REG_KEY32 + productCode + "_is1";
+            var key = OpenKey(issProdKey);
+            if (null == key)
+            {
+                issProdKey = REG_KEY64 + productCode;
+                key = OpenKey(issProdKey);
+            }
+            productName = GetDisplayName(key);
+            return GetInstallLocation(key);
+        }
+
+        public virtual string GetCoreFilePathFromInstallation(string installPath)
+        {
+            return Directory.GetFiles(installPath, fileLookup).First();
+        }
+
+        public virtual Tuple<int, int, int, int> GetVersionInfoFromFile(string filePath)
+        {
+            var version = FileVersionInfo.GetVersionInfo(filePath);
+            return Tuple.Create(version.FileMajorPart, version.FileMinorPart, version.FileBuildPart, version.FilePrivatePart);
+        }
+    }
+
+    class InstalledProduct : IInstalledProduct
+    {
+        public string ProductName { get; set; }
+        public string InstallLocation { get; set; }
+        public string VersionString { get; set; }
+        public Tuple<int, int, int, int> VersionInfo { get; private set; }
+
+        public InstalledProduct(string installLocation, InstalledProductLookUp lookUp)
+        {
+            InstallLocation = installLocation;
+            var corePath = lookUp.GetCoreFilePathFromInstallation(InstallLocation);
+            VersionInfo = lookUp.GetVersionInfoFromFile(corePath);
+            ProductName = string.Format("{0} {1}.{2}", lookUp.ProductLookUpName, VersionInfo.Item1, VersionInfo.Item2);
+            VersionString = string.Format("{0}.{1}.{2}.{3}", VersionInfo.Item1, VersionInfo.Item2, VersionInfo.Item3, VersionInfo.Item4);
+        }
+
+        public int CompareTo(object obj)
+        {
+            var product = obj as IInstalledProduct;
+            if (null == product)
+                return -100000;
+
+            var x = (VersionInfo.Item1 - product.VersionInfo.Item1);
+            if (x != 0)
+                return x*10000;
+
+            x = (VersionInfo.Item2 - product.VersionInfo.Item2);
+            if (x != 0)
+                return x*100;
+
+            return (VersionInfo.Item3 - product.VersionInfo.Item3);
+        }
+
+        public override bool Equals(object obj)
+        {
+            return CompareTo(obj) == 0;
+        }
+
+        public override int GetHashCode()
+        {
+            return VersionInfo.GetHashCode();
+        }
+    }
```

Comment:
I don't think you need to override `GetHashCode` and `Equals` if all you're doing is implementing `IComparable` for sorting.